File: factory.cpp
#include <iostream>

class Object {
public:
    virtual ~Object() = default;;

    virtual void foo() = 0;

    virtual void bar() = 0;

    virtual void sayHello(std::string_view name) = 0;
};

class ObjectA : public Object {
    ~ObjectA() override { std::cout << "ObjectA::Destructor\n"; }


    void foo() override {
        std::cout << "ObjectA::foo\n";
    }

    void bar() override {
        std::cout << "ObjectA::bar\n";
    }

    void sayHello(std::string_view name) override {
        std::cout << "Hello, " << name << " from A\n";
    }
};

class ObjectB : public Object {
    ~ObjectB() override { std::cout << "ObjectB::Destructor\n"; }

    void foo() override {
        std::cout << "ObjectB::foo\n";
    }

    void bar() override {
        std::cout << "ObjectB::bar\n";
    }

    void sayHello(std::string_view name) override {
        std::cout << "Hello, " << name << " from B\n";
    }
};

class Factory {
    virtual Object *makeObject() const = 0;

public:
    void run() {
        auto object = this->makeObject();
        object->foo();
        object->bar();
        object->sayHello("Albert");
        delete object;
    }
};


class FactoryA : public Factory {
    Object *makeObject() const override {
        return new ObjectA();
    }
};

class FactoryB : public Factory {
    Object *makeObject() const override {
        return new ObjectB();
    }
};

int main() {
    FactoryA factoryA{};
    factoryA.run();

    FactoryB factoryB{};
    factoryB.run();

    return 0;
}

File: factory.py
from __future__ import annotations
from abc import ABC, abstractmethod

class MyObject(ABC):

    @abstractmethod
    def foo(self) -> None:
        pass

    @abstractmethod
    def bar(self) -> None:
        pass

    @abstractmethod
    def say_hello(self, name: str) -> None:
        pass

class ObjectA(MyObject):

    def foo(self) -> None:
        print("ObjectA::foo")

    def bar(self) -> None:
        print("ObjectA::bar")

    def say_hello(self, name: str) -> None:
        print(f"ObjectA::say_hello {name}")

class ObjectB(MyObject):

    def foo(self) -> None:
        print("ObjectB::foo")

    def bar(self) -> None:
        print("ObjectB::bar")
    
    def say_hello(self, name: str) -> None:
        print(f"ObjectB::say_hello {name}")

class Factory(ABC):
    @abstractmethod
    def make_object(self) -> MyObject:
        pass
    
    def run(self) -> None:
        obj = self.make_object()
        obj.foo()
        obj.bar()
        obj.say_hello("Albert")

class FactoryA(Factory):
    def make_object(self) -> ObjectA:
        return ObjectA()

class FactoryB(Factory):
    def make_object(self) -> ObjectB:
        return ObjectB()

def main():
    factoryA = FactoryA()
    factoryA.run()

    factoryB = FactoryB()
    factoryB.run()

if __name__ == "__main__":
    main()
